### ISSUE TITLE
Use of a = b.clone() instead of a.clone_from(&b)

### ISSUE THREAD

**Issue by laralove143** [rust-lang/rust-clippy#9512]
### What it does

Catches the places where `clone_from` could be used instead

### Lint Name

clone_from_candidate

### Category

pedantic

### Advantage

From the [docs](https://doc.rust-lang.org/std/clone/trait.Clone.html#method.clone_from)
- `a.clone_from(&b)` is equivalent to `a = b.clone()` in functionality, but can be overridden to reuse the resources of a to avoid unnecessary allocations.

### Drawbacks

I can't think of any

### Example

```rust
let mut a = "a".to_owned();
let b = "b".to_owned();
a = b.clone();
drop(b);
```

Could be written as:

```rust
let mut a = "a".to_owned();
let b = "b".to_owned();
a.clone_from(&b);
drop(b);
```

**Comment by danielrab** [comment]
the snippet of code provided doesn't exactly work, but using `b.clone_into(&mut a);` instead of `a.clone_from(b);` does.

**Comment by Alexendoo** [comment]
It would be good if the lint could check if the type has a provided `fn clone_from`, if it's using the default from `Clone` the suggestion wouldn't change anything

**Comment by laralove143** [comment]
> the snippet of code provided doesn't exactly work, but using `b.clone_into(&mut a);` instead of `a.clone_from(b);` does.

Fixed it

**Comment by y21** [comment]
This was implemented in #12077

Output for the example
```
warning: assigning the result of `Clone::clone()` may be inefficient
 --> src/main.rs:5:5
  |
5 |     a = b.clone();
  |     ^^^^^^^^^^^^^ help: use `clone_from()`: `a.clone_from(&b)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assigning_clones
  = note: `#[warn(clippy::assigning_clones)]` on by default
```
